File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Common/SpawnMonsterAction.ts
import Async from "../../../../../Core/Async";
import DungeonManager from "../../DungeonManager";
import { OnSpawnMonsterMessage } from "../../Events/OnSpawnMonsterMessage";
import { MinionPower } from "../../Power/MinionPower";
import { AbstractMonster } from "../../Unit/Monster/AbstractMonster";
import AbstractGameAction from "../AbstractGameAction";

export class SpawnMonsterAction extends AbstractGameAction{
    public constructor(public monster : AbstractMonster,public isMinion:boolean) {
        super();
    }
    
    public async Update(): Promise<boolean> {
        if(this.Duration == 0){
            DungeonManager.MessageManager.Send(OnSpawnMonsterMessage.Id,new OnSpawnMonsterMessage(this.monster));
            this.monster.Initialize();
            if(this.isMinion){
                this.monster.AddPower(new MinionPower(),this.monster);
            }
        }
        await Async.Delay(100);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Combat/SpawnMonsterAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Dungeon;
using ModTheSpire.Event;
using ModTheSpire.Event.Ids;
using ModTheSpire.Power.Special;
using ModTheSpire.Room;
using ModTheSpire.Unit.Monster;

namespace ModTheSpire.Effect.Other
{
    public class SpawnMonsterAction : IGameAction
    {
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.SPECIAL;
        private AbstractMonster Monster;
        private bool IsMinion;
        public SpawnMonsterAction(AbstractMonster m,bool isMinion)
        {
            Monster = m;
            IsMinion = isMinion;
        }
        
        public void Init()
        {
            throw new System.NotImplementedException();
        }

        public async UniTask<bool> Update()
        {
            DungeonManager.MessageManager.Send(new OnSpawnMonsterMessage(Monster));
            if (DungeonManager.Inst.CurrentRoom is CombatRoom room)
            {
                room.Monsters.AddMonster(Monster);
                if (IsMinion)
                {
                    Monster.AddPower(new MinionPower(), Monster);
                }
            }
            await UniTask.Delay(100);
            return true;
        }
    }
}